File: Stacks-Queues/min-remove-to-make-valid-parentheses.js
/**
 * 1249. Minimum Remove to Make Valid Parentheses
 * 
 * Given a string s of '(' , ')' and lowercase English characters.
 * Your task is to remove the minimum number of parentheses ( '(' or ')', in any positions ) 
 * so that the resulting parentheses string is valid and return any valid string.
 * 
 * Formally, a parentheses string is valid if and only if:
 * It is the empty string, contains only lowercase characters, or
 * It can be written as AB (A concatenated with B), where A and B are valid strings, or
 * It can be written as (A), where A is a valid string.
 * 
 * @param {string} s
 * @return {string}
 */
var minRemoveToMakeValid = function(s) {
    const arrayString = s.split("");
    const stack = [];
    for (let i = 0; i<arrayString.length; i++) {
        // If the char is a left bracket push the index into the stack
        if (arrayString[i] === '(') {
            stack.push(i);
        } else if (arrayString[i] === ')' && stack.length) {
            // If the char is right bracket and there exists left bracket in the stack they cancel each other
            stack.pop();
        } else if (arrayString[i] === ')') {
            // If there is no left bracket in the stack and only the right bracket, remove the element.
            arrayString[i] = '';
        }
    }

    // If there are some left brackets remaining remove those elements
    while (stack.length) {
        const deleteIndex = stack.pop();
        arrayString[deleteIndex] = "";
    }

    return arrayString.join("");
    
};

File: Stacks-Queues/min-remove-to-make-valid-parentheses.py
# Minimum Remove to Make Valid Parentheses
# Given a string s of '(' , ')' and lowercase English characters.

# Your task is to remove the minimum number of parentheses ( '(' or ')', in any positions ) so that the resulting parentheses string is valid and return any valid string.

# Formally, a parentheses string is valid if and only if:

# It is the empty string, contains only lowercase characters, or
# It can be written as AB (A concatenated with B), where A and B are valid strings, or
# It can be written as (A), where A is a valid string.

class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        leftBracketStack = []
        s = list(s)
        charToRemove = []

        # Iterate through the string to identify invalid parentheses
        for i in range(len(s)):
            if s[i] == ')':
                # Check if there is a matching open parenthesis
                if leftBracketStack:
                    leftBracketStack.pop()
                else:
                    # If no matching open parenthesis, mark for removal
                    charToRemove.append(i)
            elif s[i] == '(':
                # Add the index of the open parenthesis to the stack
                leftBracketStack.append(i)
        
        # Add the remaining unmatched open parentheses to charToRemove
        charToRemove.extend(leftBracketStack)
        charToRemove.sort(reverse=True)

        # Remove marked characters from the string
        if s:
            for i in charToRemove:
                s.pop(i)
        
        return "".join(s)

# Time Complexity: O(n), where n is the length of the input string
# Space Complexity: O(n), where n is the length of the input string (used for the stack and charToRemove list)
